SofaKernel/modules/SofaHelper/src/sofa/helper/NameDecoder.h

```diff
@@ -0,0 +1,267 @@
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+
+template<class T>
+class HasGetCustomClassName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::GetCustomClassName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+//TODO(dmarchal: 01/04/2020) This is for compatibility layer, remove it after 01/01/2021
+template<class T>
+class HasGetDefaultTemplateName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::GetDefaultTemplateName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+//TODO(dmarchal: 01/04/2020) This is for compatibility layer, remove it after 01/01/2021
+template<class T>
+class HasDeprecatedTemplateName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::templateName) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+//TODO(dmarchal: 01/04/2020) This is for compatibility layer, remove it after 01/01/2021
+template<class T>
+class HasDeprecatedClassName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::className) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+template<class T>
+class HasName
+{
+    typedef char YesType[1];
+    typedef char NoType[2];
+
+    template<typename C> static YesType& test( decltype (&C::Name) );
+    template<typename C> static NoType& test(...);
+
+public:
+    enum { value = sizeof(test<T>(0)) == sizeof(YesType) };
+};
+
+template<typename T>
+std::string GetSofaTypeTemplateName(const std::string prefix="")
+{
+    if constexpr (HasName<T>::value )
+            return prefix + T::Name();
+    return "";
+}
+
+template<typename T, typename ...Ts>
+std::enable_if_t<sizeof...(Ts)!=0, std::string>
+GetSofaTypeTemplateName(const std::string prefix="")
+{
+    return GetSofaTypeTemplateName<T>(prefix) + GetSofaTypeTemplateName<Ts...>(",");
+}
+
+class SOFA_HELPER_API NameDecoder
+{
+public:
+    template <typename T>
+    struct DefaultTypeTemplateName {
+        static std::string Get() { return "MERDE"; }
+    };
+
+    template<template <typename, typename...> class C, typename T1, typename ...Ts>
+    struct DefaultTypeTemplateName<C<T1,Ts...>>{
+        static std::string Get()
+        {
+            return GetSofaTypeTemplateName<T1, Ts...>();
+        }
+    };
+
+private:
+    /// Take this function if there a className but a CustomClassName
+    template<class T, typename std::enable_if<HasGetCustomClassName<T>::value, int>::type = 0>
```

Comment:
Could you do only one method `getOverridableName()` with `constexpr if` instead of multiple SFINAE methods with `enable_if` that are simply awful to read, extend and understand? Especially if you have specialization order "priorities". Something like:

```
template<class T>
static const std::string getOverridableName() {
    if constexpr(HasGetCustomClassName<T>::value) {
        return T::GetCustomClassName();
   } else if constexpr(HasDeprecatedClassName<T>::value && !HasGetCustomClassName<T>::value) {
        T* ptr {nullptr};
        return T::className(ptr);
    }
    (...)
}
```

Reply:
You are totally right this code is dating back from the first version when I was using sfinae. I could update that part.